marimo/_runtime/runner/hooks_post_execution.py

```diff
@@ -169,6 +169,17 @@ def _store_reference_to_output(
             cell.set_output(run_result.output)
 
 
+def _store_state_reference(
+    cell: CellImpl,
+    runner: cell_runner.Runner,
+    run_result: cell_runner.RunResult,
+) -> None:
+    del run_result
+    ctx = get_context()
+    ctx.state_registry.register_scope(runner.glbls, defs=cell.defs)
+    ctx.state_registry.retain_active_states(set(runner.glbls.keys()))
```

Comment:
Can you explain why this is necessary? For UI elements we only register on creation, and that I think is sufficient. I noticed state objects are already registered on creation in `State.__init__`.

Reply:
> I already recognize this is an enormous PR, so LMK if I should just try to address the documentation and cleanup here, and tackle these issues in another PR- or try to get it right from the beginning. 

Up to you. You can tackle in another PR if you like -- just leave a comment or TODO in the source if you do.